<!DOCTYPE html><html lang="en-us"><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1"><title>Cluster Animation - ACTing for U</title><meta name="description" content="Week 8 Preliminary animation of human genome uncharacterized proteins. Each frame is an individual sequence with the nucleotides mapped individually in color. (See code below) # from oneball.py (should also copy onto cluster) of VIS 141A import os import os.path import sys import pygame from&hellip;"><meta name="generator" content="Publii Open-Source CMS for Static Site"><link rel="canonical" href="https://n-katty-kat.github.io/ACTinG-for-U/cluster-animation.html"><link rel="alternate" type="application/atom+xml" href="https://n-katty-kat.github.io/ACTinG-for-U/feed.xml" title="ACTing for U - RSS"><link rel="alternate" type="application/json" href="https://n-katty-kat.github.io/ACTinG-for-U/feed.json" title="ACTing for U - JSON"><meta property="og:title" content="Cluster Animation"><meta property="og:site_name" content="ACTing for U"><meta property="og:description" content="Week 8 Preliminary animation of human genome uncharacterized proteins. Each frame is an individual sequence with the nucleotides mapped individually in color. (See code below) # from oneball.py (should also copy onto cluster) of VIS 141A import os import os.path import sys import pygame from&hellip;"><meta property="og:url" content="https://n-katty-kat.github.io/ACTinG-for-U/cluster-animation.html"><meta property="og:type" content="article"><link rel="preload" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/dynamic/fonts/robotoflex/robotoflex.woff2" as="font" type="font/woff2" crossorigin><link rel="preload" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/dynamic/fonts/robotoslab/robotoslab.woff2" as="font" type="font/woff2" crossorigin><link rel="stylesheet" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/css/style.css?v=f6bdb60e55e38d9674e50e2698bccd14"><script type="application/ld+json">{"@context":"http://schema.org","@type":"Article","mainEntityOfPage":{"@type":"WebPage","@id":"https://n-katty-kat.github.io/ACTinG-for-U/cluster-animation.html"},"headline":"Cluster Animation","datePublished":"2025-02-03T16:29-08:00","dateModified":"2025-03-06T20:56-08:00","description":"Week 8 Preliminary animation of human genome uncharacterized proteins. Each frame is an individual sequence with the nucleotides mapped individually in color. (See code below) # from oneball.py (should also copy onto cluster) of VIS 141A import os import os.path import sys import pygame from&hellip;","author":{"@type":"Person","name":"AJ (n.kat)","url":"https://n-katty-kat.github.io/ACTinG-for-U/authors/n-kat/"},"publisher":{"@type":"Organization","name":"AJ (n.kat)"}}</script><noscript><style>img[loading] {
                    opacity: 1;
                }</style></noscript></head><body class="page-template"><header class="top js-header"><a class="logo" href="https://n-katty-kat.github.io/ACTinG-for-U/">ACTing for U</a><nav class="navbar js-navbar"><button class="navbar__toggle js-toggle" aria-label="Menu" aria-haspopup="true" aria-expanded="false"><span class="navbar__toggle-box"><span class="navbar__toggle-inner">Menu</span></span></button><ul class="navbar__menu"><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/" target="_self">IxU Reflected</a></li><li class="active-parent has-submenu"><a href="https://n-katty-kat.github.io/ACTinG-for-U/progress.html" target="_self" aria-haspopup="true">Progress</a><ul class="navbar__submenu level-2" aria-hidden="true"><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/water-environment.html" target="_self">Water Environment</a></li><li class="active"><a href="https://n-katty-kat.github.io/ACTinG-for-U/cluster-animation.html" target="_self">Cluster Animation</a></li></ul></li><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/concept-and-lineage.html" target="_self">Concept and Lineage</a></li><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/getting-technical.html" target="_self">Getting Technical</a></li><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/about-me.html" target="_self">About Me</a></li></ul></nav></header><main class="page"><article class="content"><div class="hero hero--noimage"><header class="hero__content"><div class="wrapper"><h1>Cluster Animation</h1></div></header></div><div class="entry-wrapper content__entry"><div class="post__toc"><h3>Table of Contents</h3><ul><li><a href="#mcetoc_1ildb1k5glf">Week 8</a></li><li><a href="#mcetoc_1ildb1k5glg">Week 7</a></li><li><a href="#mcetoc_1ildb1k5glg">Week 5</a></li></ul></div><h3 id="mcetoc_1ildb1k5glf">Week 8</h3><p class="align-center"></p><div class="post__iframe"><iframe loading="lazy" width="697" height="392" style="width: 697px; height: 392px;" src="https://www.youtube.com/embed/_pPPqxlxh7g?si=Lr_UxUsH6KAmk8Pa" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen="allowfullscreen"></iframe></div>Preliminary animation of human genome uncharacterized proteins. Each frame is an individual sequence with the nucleotides mapped individually in color. (See code below)<p></p><pre class="language-python"><code># from oneball.py (should also copy onto cluster) of VIS 141A
import os
import os.path
import sys
import pygame
from pygame.locals import *
import time
import random
# from mpi4py import MPI
import math
import sqlite3

# comm = MPI.COMM_WORLD
# rank = comm.Get_rank()
# size = comm.Get_size()

# set window position
os.environ['SDL_VIDEO_WINDOW_POS'] = "0,0"
# get this display
#print(dict(os.environ))
os.environ['DISPLAY'] = ':0.0'
# with caffeine installed (sudo apt install caffeine) we can wake the screen
# os.system('caffeinate sleep 1') # passes caffeinate out to shell to wake screensaver

# init clock and display
clock = pygame.time.Clock()
pygame.display.init()
pygame.mouse.set_visible(False)

# get the screen hight and width
disp_info = pygame.display.Info()
width = disp_info.current_w
height = disp_info.current_h
screen_size = (width,height) # adjust screen size manually is monitor is not the same dimensions as the main monitor
# universe_size = (width*2, height*2)
screen = pygame.display.set_mode((screen_size), pygame.NOFRAME)

# sqlite search cross with csv --&gt; get list of uncharacterized protein codes, find each of those in sqlite database
    # wondering if i should just put the protein names into sqlite db as well to better describe the protein and just search that way
    # updated database: genome_human (gene VARCHAR(14), protein VARCHAR(132), sequence VARCHAR(107976), changed_sequence INTEGER)
        # defaults changed_sequence to 0 (represents false), changes to 1 if serial data being read is giving into a change to the sequence (duplicate database or dataset to read)

connection = sqlite3.connect('genome_human.db') # change to full path name?
cursor = connection.cursor()

cursor.execute("SELECT sequence FROM genome_human WHERE protein LIKE '%uncharacterized%'")
# SELECT * will return [(gene, protein, sequence),...]
    # () &lt;- tuple so even if 1 its still (value,) at least
    # obtain thru tuple indexing: seq1 = var[0][0] (1st list item, index 0 of tuple)
var1 = cursor.fetchall()
# cursor.execute("SELECT protein FROM genome_human WHERE protein LIKE 'uncharacterized%'")
# var2 = cursor.fetchall()
# print(len(var1)) #1638 (for protein)
# print(len(var2)) #1564 (for protein)

# check proteins that have augment before "uncharacterized"
# for p in var1:
#     if p not in var2:
#         print(p)
# described uncharacterized:
#     ('LOW QUALITY PROTEIN: uncharacterized protein',)
#     ('putative uncharacterized protein C3orf49',)
# do i want to include those above or even make them smth different based on being putative or low quality?
# print(var)
framerate = 24
runtime = 10 #seconds

def nucleotide_color(nucleotide):
    if nucleotide == "A":
        return 50
    elif nucleotide == "C":
        return 100
    elif nucleotide == "T":
        return 150
    elif nucleotide == "G":
        return 200
    else:
        return 255

def sine_wave(shift,sequence):
    len_sequence = len(sequence)
    for i in range(height):
        # sine transformations -&gt; bsin(ax-w)+h (b increases amplitude, a gives inverse frequency increase, w shifts horizontally, h shifts vertically)
        # pygame.draw.line(screen,(50,100,255),(0,i),((width/2)+((width/2)*math.sin((height/4)*(i-shift))),i)) 
            # moves down the screen
        # pygame.draw.line(screen,(50,100,255),(0,i),((width/2)+((width/2)*math.sin((height/4)*i)),i))
            # this (width/2)+((width/2)*math.sin((height/4)*i)) did NOT do what i expected at all but its GORGEOUS
        if i &gt;= len_sequence:
            pygame.draw.line(screen,(255,255,255),(0,i),((width/2)+((width/2)*math.sin((height/4)*i)),i))
        else:
            pygame.draw.line(screen,(50,100,nucleotide_color(sequence[i])),(0,i),((width/2)+((width/2)*math.sin((height/4)*i)),i))
    return

protein_index = 0
# time to run
# timeToRun = framerate*runtime
timeToRun = len(var1)

for f in range(timeToRun):
    # screen.fill((50,30,255))
    screen.fill((0,0,0))
    sine_wave(f,var1[protein_index][0]) #shift later to go thru each sequence, this would be per frame but i could try other methods later
    pygame.display.update()
    protein_index = protein_index + 1
    if protein_index &gt; len(var1):
        protein_index = 0
    clock.tick(framerate)

connection.close()</code></pre><h3 id="mcetoc_1ildb1k5glg">Week 7</h3><div class="gallery-wrapper"><div class="gallery" data-is-empty="false" data-translation="Add images" data-columns="3"><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250226_124440.jpg" data-size="2801x2172"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250226_124440-thumbnail.jpg" alt="" width="720" height="558"></a><figcaption>2x4 screen display setup with vertical screen placement</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups2.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups2-thumbnail.png" alt="" width="720" height="405"></a><figcaption>2x4 vertical screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups-thumbnail.png" alt="" width="720" height="405"></a><figcaption>4x2 horizontal screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups1.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups1-thumbnail.png" alt="" width="720" height="405"></a><figcaption>3x2 horizontal screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups3.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups3-thumbnail.png" alt="" width="720" height="405"></a><figcaption>2x4 horizontal screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups6.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups6-thumbnail.png" alt="" width="720" height="405"></a><figcaption>Split 2x2, 2x2 screen displays</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups5.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups5-thumbnail.png" alt="" width="720" height="405"></a><figcaption>1x8 vertical screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups4.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups4-thumbnail.png" alt="" width="720" height="405"></a><figcaption>4x2 vertical screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups7.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups7-thumbnail.png" alt="" width="720" height="405"></a><figcaption>2 layer shifted 8-screen display</figcaption></figure></div></div><p>Rearranging screen display for cluster: 3x3 horizontal screens to 2x4 vertical screens. Intended to improve scaling and utilization of wall space.</p><p>Fixed genome .fna file reading: Wrote out gene ID, protein name, and sequence to SQLite database. (See code below)</p><p>Fixed PySerial communication and ran sample code to print out simple Serial output from Arduino:</p><ol style="list-style-type: lower-roman;"><li><a href="https://roboticsbackend.com/raspberry-pi-arduino-serial-communication/">https://roboticsbackend.com/raspberry-pi-arduino-serial-communication/</a></li></ol><pre class="language-python"><code># reading the human genome fna file

# code ref from ch 19 https://coding4medicine.com/backup/Python/reading-fasta-files.html
import sys
import time
import sqlite3

f=open("ncbi_dataset/data/GCF_000001405.40/cds_from_genomic.fna", "r") # open file with the genome data

connection = sqlite3.connect('genome_human.db') # change to full path name?
cursor = connection.cursor()

cursor.execute('CREATE TABLE genome_human (gene VARCHAR(14), protein VARCHAR(132), sequence VARCHAR(107976), changed_sequence INTEGER)') #determine the appropriate varchar max for each of these by reading through the file
# https://dba.stackexchange.com/questions/190674/can-a-database-table-not-have-a-primary-key 
    # use rowid to see primary key
# changed_sequence is boolean as integer (0 false, 1 true) --&gt; indicates whether the sequence item has been changed or not by external adjustments to its nucleotides

connection.commit() # commit to creating the file before trying to insert into it

for line in f:
    # read over each line of the file
    if(line.startswith("&gt;")):
        if(new_gene):
            new_gene = False
            sequence_list = sequence_string.split("\n") # remove the \n from btwn each sequence subpart from the fna format and any blank lines
            sequence_string = "" #clear out sequence so that it doesnt concatenate old sequences w new sequences
            sequence_output = "".join(sequence_list) # join together that list to form one continuous sequence line

            cursor.execute('INSERT INTO genome_human VALUES (?,?,?,0);', (gene_id, protein_name, sequence_output))

        identity = line.split(" [") # split the elements of each into 1c1, gene], db], pro], proid], loc], gbkey] (since protein= can have spaces in btwn)
            # gene always gene=###] --&gt; same way to obtain as protein_id with splice or split and cut 5 (after =) to length
            # protein_id will always be protein_id=#####] so can obtain number as split at = and get num minus ending bracket or splice index 11 to length (to keep out bracket)
        # get gene
        for i in identity:
            if(i.startswith("gene=")):
                gene_id = i
                break
        gene_id_len = len(gene_id)
        gene_id = gene_id[5:(gene_id_len-1)] # cuts out "gene=" and "]" ; alternately can put len("gene=") in place of 6

        for i in identity:
            if(i.startswith("protein=")):
                protein_name = i
                break
        protein_name_len = len(protein_name)
        protein_name = protein_name[len("protein="):(protein_name_len-1)]
        
    else:
        new_gene = True
        sequence_string = sequence_string + line # once starting a new gene, the string will be split at "\n" and pieced back together as a string

print("Done")
f.close()
connection.commit()
cursor.close()
connection.close()</code></pre><h3 id="mcetoc_1ildb1k5glg">Week 5</h3><p>Researching FASTA format of human genome data:</p><ol style="list-style-type: lower-roman;"><li><a href="https://coding4medicine.com/backup/Python/reading-fasta-files.html">https://coding4medicine.com/backup/Python/reading-fasta-files.html</a></li><li><a href="https://en.wikipedia.org/wiki/FASTA_format">https://en.wikipedia.org/wiki/FASTA_format</a></li><li><a href="https://www.ncbi.nlm.nih.gov/datasets/taxonomy/9606/">https://www.ncbi.nlm.nih.gov/datasets/taxonomy/9606/</a></li></ol><p>Testing reading out genome data in Python:</p><ol style="list-style-type: lower-roman;"><li>Issue: CSV file ended up being over 9GB!! (from a 2GB FASTA .fna file)</li><li>Resolution: The variable holding each sequence was not being cleared for each new sequence and continuously concatenated the sequences (memory leak...).</li></ol><p>Current issue: PySerial not installing on Raspberry Pi 5 (Spoiler Alert: It was already pre-installed)</p></div></article></main><footer class="footer"><div class="wrapper"><div class="footer__copyright"><p>Jett 2025 - Powered by Publii</p></div><button onclick="backToTopFunction()" id="backToTop" class="footer__bttop" aria-label="Back to top" title="Back to top"><svg width="20" height="20"><use xlink:href="https://n-katty-kat.github.io/ACTinG-for-U/assets/svg/svg-map.svg#toparrow"/></svg></button></div></footer><script defer="defer" src="https://n-katty-kat.github.io/ACTinG-for-U/assets/js/scripts.min.js?v=700105c316933a8202041b6415abb233"></script><script>window.publiiThemeMenuConfig={mobileMenuMode:'sidebar',animationSpeed:300,submenuWidth: 'auto',doubleClickTime:500,mobileMenuExpandableSubmenus:true,relatedContainerForOverlayMenuSelector:'.top'};</script><script>var images = document.querySelectorAll('img[loading]');
        for (var i = 0; i < images.length; i++) {
            if (images[i].complete) {
                images[i].classList.add('is-loaded');
            } else {
                images[i].addEventListener('load', function () {
                    this.classList.add('is-loaded');
                }, false);
            }
        }</script></body></html>