<!DOCTYPE html><html lang="en-us"><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,initial-scale=1,user-scalable=no"><title>Cluster Animation - ACTing for U</title><meta name="robots" content="noindex,nofollow"><meta name="generator" content="Publii Open-Source CMS for Static Site"><link rel="alternate" type="application/atom+xml" href="https://n-katty-kat.github.io/ACTinG-for-U/feed.xml" title="ACTing for U - RSS"><link rel="alternate" type="application/json" href="https://n-katty-kat.github.io/ACTinG-for-U/feed.json" title="ACTing for U - JSON"><meta property="og:title" content="Cluster Animation"><meta property="og:site_name" content="ACTing for U"><meta property="og:description" content="Week 18 Screen glare from open door Full room lighting Dim room lighting Final full-screen highlight to identify gene macropixel being changed Test&hellip;"><meta property="og:url" content="https://n-katty-kat.github.io/ACTinG-for-U/cluster-animation.html"><meta property="og:type" content="article"><link rel="shortcut icon" href="https://n-katty-kat.github.io/ACTinG-for-U/media/website/favicon-2.png" type="image/png"><link rel="preload" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/dynamic/fonts/redhatdisplay/redhatdisplay.woff2" as="font" type="font/woff2" crossorigin><link rel="preload" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/dynamic/fonts/redhatdisplay/redhatdisplay-italic.woff2" as="font" type="font/woff2" crossorigin><link rel="preload" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/dynamic/fonts/instrumentsans/instrumentsans.woff2" as="font" type="font/woff2" crossorigin><link rel="preload" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/dynamic/fonts/instrumentsans/instrumentsans-italic.woff2" as="font" type="font/woff2" crossorigin><link rel="stylesheet" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/css/fontawesome-all.min.css?v=85514f933f9e0b82460af63f1a403fa5"><link rel="stylesheet" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/css/style.css?v=dcdf2fbed1c109bee75b2dac8027b0c7"><noscript><link rel="stylesheet" href="https://n-katty-kat.github.io/ACTinG-for-U/assets/css/noscript.css?v=efa867a99f5064d6729e4dc2008ad50b"></noscript><script type="application/ld+json">{"@context":"http://schema.org","@type":"Article","mainEntityOfPage":{"@type":"WebPage","@id":"https://n-katty-kat.github.io/ACTinG-for-U/cluster-animation.html"},"headline":"Cluster Animation","datePublished":"2025-02-03T16:29-08:00","dateModified":"2025-06-11T19:18-07:00","description":"Week 18 Screen glare from open door Full room lighting Dim room lighting Final full-screen highlight to identify gene macropixel being changed Test&hellip;","author":{"@type":"Person","name":"AJ (n.kat)","url":"https://n-katty-kat.github.io/ACTinG-for-U/authors/n-kat/"},"publisher":{"@type":"Organization","name":"AJ (n.kat)"}}</script><style>#wrapper > .bg {
               background-image: url(https://n-katty-kat.github.io/ACTinG-for-U/assets/images/overlay.png), linear-gradient(0deg, rgba(0, 0, 0, 0.1), rgba(0, 0, 0, 0.1)), url(https://n-katty-kat.github.io/ACTinG-for-U/media/website/20250610_1502171.jpg);
           }</style><noscript><style>img[loading] {
                    opacity: 1;
                }</style></noscript></head><body class="is-preload page-template"><div id="wrapper"><header id="header"><a class="logo" href="https://n-katty-kat.github.io/ACTinG-for-U/">ACTing for U</a></header><nav id="nav"><ul class="links"><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/" target="_self">ACTinG for U</a></li><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/lineage.html" target="_self">Lineage</a></li><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/behavior.html" target="_self">Behavior</a></li><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/timeline.html" target="_self">Timeline</a></li><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/development.html" target="_self">Development</a></li><li><a href="https://n-katty-kat.github.io/ACTinG-for-U/about-me.html" target="_self">About Me</a></li></ul></nav><main id="main"><article class="post"><header class="major"><h1>Cluster Animation</h1><p class="post__inner"></p></header><div class="post__inner post__entry"><div class="post__toc"><h3>Table of Contents</h3><ul><li><a href="#mcetoc_1is06ed0o3t">Week 18</a></li><li><a href="#mcetoc_1is06ed0o3u">Week 17</a></li><li><a href="#mcetoc_1is06ed0o3v">Week 16</a></li><li><a href="#mcetoc_1is06ed0o40">Week 15</a></li><li><a href="#mcetoc_1ipqkcu2p4s">Week 13</a></li><li><a href="#mcetoc_1ipqgrah84o">Week 12</a></li><li><a href="#mcetoc_1ipqc1fps41">Week 11</a></li><li><a href="#mcetoc_1ipqc1fps42">Week 9</a></li><li><a href="#mcetoc_1ildb1k5glf">Week 8</a></li><li><a href="#mcetoc_1ildb1k5glg">Week 7</a></li><li><a href="#mcetoc_1ildb1k5glg">Week 5</a></li></ul></div><h3 id="mcetoc_1is06ed0o3t">Week 18</h3><div class="gallery-wrapper"><div class="gallery" data-is-empty="false" data-translation="Add images" data-columns="3"><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250521_104849.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250521_104849-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Screen glare from open door</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250522_160537.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250522_160537-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Full room lighting</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250522_160546.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250522_160546-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Dim room lighting</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250522_160526.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250522_160526-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Final full-screen highlight to identify gene macropixel being changed</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250521_114342.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250521_114342-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Test bullseye to better identify gene macropixel being changed</figcaption></figure></div></div><p class="align-center"></p><div class="post__iframe"><iframe loading="lazy" width="688" height="387" style="width: 688px; height: 387px;" src="https://www.youtube-nocookie.com/embed/T8RMuUOiL9I?si=0LGr_1KAFm4fyzu1" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen="allowfullscreen"></iframe></div>Animation updated visual running with water system<p></p><ul><li style="font-weight: 400;" aria-level="1"><span style="font-weight: 400;">Updated animation visuals to better emphasize the gene macropixel getting changed and the color change</span></li><li style="font-weight: 400;" aria-level="1"><span style="font-weight: 400;">Adjusted the max number of bases in a sequence that can be changed by any one data input</span></li><li style="font-weight: 400;" aria-level="1"><span style="font-weight: 400;">Testing complete program (animation and water system)</span><ul><li style="font-weight: 400;" aria-level="2"><span style="font-weight: 400;">Fixed screens going out of sync due to time delays between Arduino data input and animation pauses</span></li><li style="font-weight: 400;" aria-level="2"><span style="font-weight: 400;">Plan to adjust initial main tank water level to be above middle level slightly (to ensure cyan dye does not drain before other behaviors can occur)</span></li><li style="font-weight: 400;" aria-level="2"><span style="font-weight: 400;">Note that starting the animation program causes the Arduino to reset.</span></li></ul></li></ul><h3 id="mcetoc_1is06ed0o3u">Week 17</h3><div class="gallery-wrapper"><div class="gallery" data-is-empty="false" data-translation="Add images" data-columns="3"><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_151233.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_151233-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>White background visibility test</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_151444.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_151444-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Light gray background visibility test</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_151320.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_151320-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Black background visibility test</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152152.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152152-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Arduino connected to raspberry pi by serial port</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152904.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152904-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Gene grid initial</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152909.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152909-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Isolated gene (row 2, column 3 monitor)</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152911.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152911-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Isolated gene updated color (row 2, column 3 monitor)</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152913.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152913-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Gene grid with updated gene color</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152525.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152525-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Gene grid initial close-up</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152529.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152529-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Isolated gene close-up</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152530.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152530-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Isolated gene updated color close-up</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152533.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250514_152533-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Gene grid with updated gene color close-up</figcaption></figure></div></div><p class="align-center"></p><div class="post__iframe"><iframe loading="lazy" width="693" height="389" style="width: 693px; height: 389px;" src="https://www.youtube-nocookie.com/embed/bEgJOYFPcbY?si=R2J4-BWjJNA5LeAc" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen="allowfullscreen"></iframe></div>Animation with randomized gene selection<p></p><ul><li>Running on 8-screen cluster</li></ul><pre class="language-python"><code>num_of_processes = 9</code></pre><ul><li>Randomized gene that gets updated for each data send</li></ul><pre class="language-python"><code>gene_rowid = random.randint(1, len(sequence_list)) #rank 0</code></pre><ul><li>Testing different animation background colors and room lighting for visibility (sticking with dark/neutral gray)</li></ul><h3 id="mcetoc_1is06ed0o3v">Week 16</h3><p class="align-center"></p><div class="post__iframe"><iframe loading="lazy" width="693" height="388" style="width: 693px; height: 388px;" src="https://www.youtube-nocookie.com/embed/OZ8CeDM4n00?si=bOr0seif6qz9LxPB" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen="allowfullscreen"></iframe></div>Animation Composition Adjustment<p></p><ul><li>Gene macropixel grid across full screen space (2x4 monitor set-up), centered, not subdivided per screen.</li></ul><pre class="language-python"><code>if rank != 0:
    #...#
    # create base pixel screen (separate surface)
    base_genome_surface = pygame.Surface(universe_size)
    base_genome_surface.fill(bg_color)
    for row in range(universe_grid_y): # left to right, top down ordering of genes from dataset
        for col in range(universe_grid_x):
            color_id = color_id + 1
            if color_id &gt;= len(sequence_colors):
                color_id = color_id_start
            try:
                pygame.draw.circle(base_genome_surface, sequence_colors[color_id], (upper_left_corner[0]+(pixel_width//2)+(pixel_width*col), upper_left_corner[1]+(pixel_height//2)+(pixel_height*row)), pixel_radius)
            except:
                print("Exception: color_id set out of bounds of sequence_colors at index", color_id)


    # create screen/surface for showing just the updated element
    update_gene_surface = pygame.Surface(universe_size)
    update_gene_surface.fill(bg_color)</code></pre><ul><li>Periodic constant Arduino data sent by serial into animation, including tweaks for rate of sending due to difference between data sending speed and animation updating speed.</li></ul><pre class="language-python"><code>while True: # rank 0, code based on https://roboticsbackend.com/raspberry-pi-arduino-serial-communication/
        if ser.in_waiting &gt; 0: # according to the arduino sends, should occur at the same rate as the arduino timer (5s, maybe make greater)
            line = ser.readline().decode('utf-8').rstrip()
            print(line)
            arduino_data = line.split(",") # ["#", "#", "#", "\n"]</code></pre><pre class="language-cpp"><code>// from SafeString V3 library https://www.forward.com.au/pfod/ArduinoProgramming/TimingDelaysInArduino.html#examples
#include "loopTimer.h"
#include "millisDelay.h"
#include "BufferedOutput.h"
createBufferedOutput(bufferedOut, 80, DROP_UNTIL_EMPTY);
millisDelay printOutTimer;


void setup() {
  Serial.begin(9600);
  bufferedOut.connect(Serial);
  printOutTimer.start(1000*10); // every 10 sec
}

void loop() {
  bufferedOut.nextByteOut();
  if(printOutTimer.justFinished()){
    printOutTimer.repeat();
    bufferedOut.print("400");
    bufferedOut.print(",");
    bufferedOut.print(220);
    bufferedOut.print(",");
    bufferedOut.print("600");
    bufferedOut.print(",");
    bufferedOut.println();
  }
}</code></pre><h3 id="mcetoc_1is06ed0o40">Week 15</h3><div class="gallery-wrapper"><div class="gallery" data-is-empty="false" data-translation="Add images" data-columns="3"><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/wk5note.jpg" data-size="1668x2157"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/wk5note-thumbnail.jpg" alt="" width="720" height="931"></a><figcaption>Notes for gridding gene macropixels</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/animationprocess.jpg" data-size="2107x1363"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/animationprocess-thumbnail.jpg" alt="" width="720" height="466"></a><figcaption>Animation flow</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250502_125500.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250502_125500-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Animation on 4-screen cluster</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250502_125539.jpg" data-size="4000x3000"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250502_125539-thumbnail.jpg" alt="" width="720" height="540"></a><figcaption>Animation close-up on one screen</figcaption></figure></div></div><p class="align-center"></p><div class="post__iframe"><iframe loading="lazy" width="689" height="387" style="width: 689px; height: 387px;" src="https://www.youtube-nocookie.com/embed/fydO9xu72Xk?si=ocdLEqBbWxuDC8xj" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen="allowfullscreen"></iframe></div>Animation Test on Cluster without serial data<p></p><ul><li>Mapped sensor data ranges into colors and bases that directly replace bases in the given gene sequence of the genome data.</li></ul><pre class="language-python"><code>m_instances = int(np.interp(arduino_data[0], [0, 3000], [0, (len(gene_edit)-1)//3])) # number of bases to replace with M, direct map higher val = more bases
c_instances = int(np.interp(arduino_data[1], [120, 362], [0, (len(gene_edit)-1)//3])) # number of bases to replace with C, map lower water level w more bases (more dye added)
y_instances = int(np.interp(arduino_data[2], [0, 1023], [0, (len(gene_edit)-1)//3])) # number of bases to replace with Y, map only above threshold values; use if statement to make negatives into 0
print("# of bases to change m c y:",m_instances, c_instances, y_instances)
# get nonrepeating indices in sequence to change
replace_indices = random.sample(range(len(gene_edit)), k=(m_instances+c_instances+y_instances)) # gets nonrepeating set (equal to total bases to be replaced) of indices in sequence to change num
## https://stackoverflow.com/questions/1228299/changing-a-character-in-a-string/22149018#22149018 -&gt; rec (for speed), splice a replacement in immutable string
# change the base at the specified index
for m in range(m_instances): # go over the first m_instances # of indices to replace, remove them each iteration to cont. use 0 index of sample list
     gene_edit = gene_edit[:replace_indices[0]] + "M" + gene_edit[replace_indices[0]+1:] # splices around the index to change (i), and concatenates the new "base" in between
     replace_indices.pop(0)
for c in range(c_instances): # go over the 1st(next) c_instances # of indices to replace, remove them each iteration to cont. use 0 index of sample list
     gene_edit = gene_edit[:replace_indices[0]] + "N" + gene_edit[replace_indices[0]+1:] # splices around the index to change (i), and concatenates the new "base" in between
     replace_indices.pop(0)
for y in range(y_instances): # go over the 1st(next) y_instances # of indices to replace, remove them each iteration to cont. use 0 index of sample list
     gene_edit = gene_edit[:replace_indices[0]] + "Y" + gene_edit[replace_indices[0]+1:] # splices around the index to change (i), and concatenates the new "base" in between
     replace_indices.pop(0)</code></pre><ul><li>Divided portion of genes to each screen (approximately 18k gene macropixels per screen).<ul><li>Ideating how to divide genes continously across all screens (instead of assigning a specific subset to individual screens).</li></ul></li></ul><pre class="language-python"><code>color_id_start = (grid_x*grid_y)*(rank-1) # set the index in the list of gene sequence colors for the given screen to start on (rank)
color_id = color_id_start</code></pre><ul><li>Scattered gene sequences across cluster nodes to process gene colors faster.</li></ul><pre class="language-python"><code>if rank == 0:
    #.....#
    num_of_processes = 4 # CHANGE WHEN PROGRAM IS ADJUSTED FOR NUM OF PROCESSES IN MPIRUN
    split_gap = len(sequence_list) // num_of_processes
    sequence_list_scatter = []
    for i in range(num_of_processes): # 0 1 2 3 4 5 6 7 8
        if i != (num_of_processes-1):
            sequence_list_scatter.append(sequence_list[split_gap*i:(split_gap*i)+split_gap])
        else:
            sequence_list_scatter.append(sequence_list[split_gap*i:]) # to end


# scatter a 2d list of num_of_processes length
print("rank", rank, "waiting for sequence_list scatter")
sequence_list_scatter = comm.scatter(sequence_list_scatter, root=0)
print("rank:", rank, "received portion of sequence_list of length", len(sequence_list_scatter))</code></pre><ul><li>Using MPI broadcast function and sleep function to control speed of screen changes in animation.</li></ul><pre class="language-python"><code>print("new color:", sequence_colors[gene_rowid-1])
print("rank", rank, "sending new sequence_colors to ani")
comm.bcast(sequence_colors, root=0) # send out list again (could alternatively send out the sequence alone + rowid but idk if thats any faster)
# ^ at this point, the animation updates color for a moment on black screen still, then returns to normal full display; and rpi5 waits for new serial again
print("rank", rank, "sent new sequence_colors to ani")
</code></pre><ul><li>Animation running while excluding the primary node from showing animation.<ul><li>8-screen cluster will have the primary node without a screen, so the command separates it from the animation and makes it only handle data processing.</li></ul></li></ul><pre class="language-bash"><code>&gt;mpirun --host node1,node2:2,node3:2,node4:2 python3 gene_animation.py</code></pre><h3 id="mcetoc_1ipqkcu2p4s">Week 13</h3><div class="gallery-wrapper"><div class="gallery" data-is-empty="false" data-translation="Add images" data-columns="3"><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/Screenshot-914.png" data-size="2400x1600"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/Screenshot-914-thumbnail.png" alt="" width="720" height="480"></a><figcaption>Sample gene sequence color generation.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/Screenshot-915.png" data-size="2400x1600"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/Screenshot-915-thumbnail.png" alt="" width="720" height="480"></a><figcaption>Sample gene sequence color generation.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/Screenshot-2025-04-20-172237.png" data-size="1641x761"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/Screenshot-2025-04-20-172237-thumbnail.png" alt="" width="720" height="334"></a><figcaption>Planned order of animation for MPI and adjusting database directly.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p6.jpg" data-size="3435x2678"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p6-thumbnail.jpg" alt="" width="720" height="561"></a><figcaption>Notes on color mapping page 6.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p7.jpg" data-size="3482x2738"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p7-thumbnail.jpg" alt="" width="720" height="566"></a><figcaption>Notes on updating gene data page 7.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p8.jpg" data-size="3441x2670"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p8-thumbnail.jpg" alt="" width="720" height="559"></a><figcaption>Notes on updating gene data page 8.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p9.jpg" data-size="3517x2776"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p9-thumbnail.jpg" alt="" width="720" height="568"></a><figcaption>Notes on MPI code outline for updating animation page 9.</figcaption></figure></div></div><p>Tested broadcasting lists with MPI as potential format for distributing genome and Arduino data across all 5 computers.</p><p>Function for generating color of a gene sequence:</p><pre class="language-python"><code>def sequence_color(sequence):
    # generates the sequence color based on % of certain nucleotides in the sequence
    len_sequence = len(sequence)
    red_count = 0
    green_count = 0
    blue_count = 0
    for n in sequence:
        if n == "A":
            red_count += 6
        elif n == "C":
            green_count += 6
        elif n == "G":
            blue_count += 6
        elif n == "T":
            red_count += 2
            green_count += 2
            blue_count += 2
        elif n == "M":
            red_count += 3
            blue_count += 3
        elif n == "N": # cyaN (since c is already used by genome)
            green_count += 3
            blue_count += 3
        elif n == "Y":
            red_count += 3
            green_count += 3
        else:
            pass # not one of the bases, could be U? otherwise an error
    # count of each nucleotide is worth 6 total so need to scale up sequence to 6 to get proper percentage i think..
        # percentage of each color to be part of the RGB color of the overall sequence
    red_per = red_count / (len_sequence*6)
    green_per = green_count / (len_sequence*6)
    blue_per = blue_count / (len_sequence*6)
    # print(red_per, green_per, blue_per)

    # actual color
    red_val = 255*red_per //1
    green_val = 255*green_per //1
    blue_val = 255*blue_per //1

    # return color
    # print(red_val, green_val, blue_val)
    return (red_val, green_val, blue_val)</code></pre><h3 id="mcetoc_1ipqgrah84o">Week 12</h3><div class="gallery-wrapper"><div class="gallery" data-is-empty="false" data-translation="Add images" data-columns="3"><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p1.jpg" data-size="3668x2826"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p1-thumbnail.jpg" alt="" width="720" height="555"></a><figcaption>Notes on new animation visuals page 1.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p2.jpg" data-size="3483x2722"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p2-thumbnail.jpg" alt="" width="720" height="563"></a><figcaption>Notes on total number of pixels and number of sequences in data page 2.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p3.jpg" data-size="3521x2714"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p3-thumbnail.jpg" alt="" width="720" height="555"></a><figcaption>Notes on gridding screen space for individual gene sequences page 3.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p4.jpg" data-size="3613x2784"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p4-thumbnail.jpg" alt="" width="720" height="555"></a><figcaption>Notes on new animation visuals page 4.</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p5.jpg" data-size="3380x2645"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/p5-thumbnail.jpg" alt="" width="720" height="563"></a><figcaption>Notes on new visual mappings for each gene page 5.</figcaption></figure></div></div><p>Concept notes on new animation visuals.</p><h3 id="mcetoc_1ipqc1fps41">Week 11</h3><p>Using PySerial to communicate Arduino's sensor data to Raspberry Pi program.</p><ul><li>Reference for PySerial: <a>https://roboticsbackend.com/raspberry-pi-arduino-serial-communication/</a></li></ul><p>Confirming that BufferedOutput library can be used to send data across serial port in place of Serial (Run PySerial program from roboticsbackend):</p><pre class="language-cpp"><code>// from SafeString V3 library https://www.forward.com.au/pfod/ArduinoProgramming/TimingDelaysInArduino.html#examples
#include "loopTimer.h"
#include "millisDelay.h"
#include "BufferedOutput.h"
createBufferedOutput(bufferedOut, 80, DROP_UNTIL_EMPTY);
millisDelay printOutTimer;

void setup() {
  Serial.begin(9600);
  bufferedOut.connect(Serial);
  printOutTimer.start(5000); // every 5 sec
}

void loop() {
  bufferedOut.nextByteOut();
  if(printOutTimer.justFinished()){ // print out the millis every 5 seconds
    printOutTimer.repeat();
    bufferedOut.println(millis());
  }
}</code></pre><p>Send Arduino sensor data (through the global behavior variables) to the Raspberry Pi in the main code (snippet) [Removed extreme cases created in Week 10 due to errors making it impossible to properly test PySerial otherwise, see Week 10 Water Environment notes]:</p><pre class="language-cpp"><code>void printData(){
  if(printOutTimer.justFinished()){
    printOutTimer.repeat();
    bufferedOut.print(socialize);
    bufferedOut.print(",");
    bufferedOut.print(hydrate);
    bufferedOut.print(",");
    bufferedOut.print(create);
    bufferedOut.print(",");
    bufferedOut.println();
  }
}</code></pre><div class="gallery-wrapper"><div class="gallery" data-is-empty="false" data-translation="Add images" data-columns="3"><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/Screenshot-2025-04-02-154537.png" data-size="320x577"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/Screenshot-2025-04-02-154537-thumbnail.png" alt="" width="320" height="577"></a><figcaption>Serial output.</figcaption></figure></div></div><p> </p><h3 id="mcetoc_1ipqc1fps42">Week 9</h3><p class="align-center"></p><div class="post__iframe"><iframe loading="lazy" width="697" height="392" style="width: 697px; height: 392px;" src="https://www.youtube-nocookie.com/embed/1j-5xVTHgHQ?si=qx4U733GNq7ozzXP&amp;start=75" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen="allowfullscreen"></iframe></div>WIP Genome Animation on 8-screen cluster set up<p></p><pre class="language-python"><code>import os
import os.path
import sys
import pygame
from pygame.locals import *
import time
import random
from mpi4py import MPI
import math
import sqlite3

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

# set window position
os.environ['SDL_VIDEO_WINDOW_POS'] = "0,0"
# get this display
#print(dict(os.environ))
os.environ['DISPLAY'] = ':0.0'
# with caffeine installed (sudo apt install caffeine) we can wake the screen
# os.system('caffeinate sleep 1') # passes caffeinate out to shell to wake screensaver

# init clock and display
clock = pygame.time.Clock()
pygame.display.init()
pygame.mouse.set_visible(False)

# get the screen hight and width
disp_info = pygame.display.Info()
width = disp_info.current_w
height = disp_info.current_h
screen_size = (width,height) # adjust screen size manually is monitor is not the same dimensions as the main monitor
universe_size = (width*2, height*2)
screen = pygame.display.set_mode((screen_size), pygame.FULLSCREEN)

connection = sqlite3.connect('genome_human.db') # change to full path name?
cursor = connection.cursor()

cursor.execute("SELECT sequence FROM genome_human WHERE protein LIKE '%uncharacterized%'")
# SELECT * will return [(gene, protein, sequence),...]
    # () &lt;- tuple so even if 1 its still (value,) at least
    # obtain thru tuple indexing: seq1 = var[0][0] (1st list item, index 0 of tuple)
var1 = cursor.fetchall()
connection.close()

def nucleotide_color(nucleotide):
    if nucleotide == "A":
        return (255,0,0) #red
    elif nucleotide == "C":
        return (0,255,0) #green
    elif nucleotide == "T":
        return (0,0,255) #blue
    elif nucleotide == "G":
        return (100,0,255) #purple?
    else:
        return (0,0,0)

def sine_wave(shift,sequence):
    len_sequence = len(sequence)
    for i in range(height):
        # sine transformations -&gt; bsin(ax-w)+h (b increases amplitude, a gives inverse frequency increase, w shifts horizontally, h shifts vertically)
        # pygame.draw.line(screen,(50,100,255),(0,i),((width/2)+((width/2)*math.sin((height/4)*(i-shift))),i))
            # moves down the screen
        # pygame.draw.line(screen,(50,100,255),(0,i),((width/2)+((width/2)*math.sin((height/4)*i)),i))
            # this (width/2)+((width/2)*math.sin((height/4)*i)) did NOT do what i expected at all but its GORGEOUS
        if i &gt;= len_sequence:
            pygame.draw.line(screen,(255,255,255),(0,i),(int((width/2)+((width/2)*math.sin((height/4)*i))),i))
        else:
            pygame.draw.line(screen,nucleotide_color(sequence[i]),(0,i),(int((width/2)+((width/2)*math.sin((height/4)*i))),i))
    return

framerate = 30
runtime = 10 #seconds
protein_index = 0
# time to run
# timeToRun = framerate*runtime
timeToRun = len(var1)

for f in range(timeToRun):
    # screen.fill((50,30,255))
    screen.fill((0,0,0))
    sine_wave(f,var1[protein_index][0]) #shift later to go thru each sequence, this would be per frame but i could try other methods later
    protein_index = protein_index + 1
    if protein_index &gt; len(var1):
        protein_index = 0
    comm.Barrier()
    pygame.display.update()
    clock.tick(framerate)</code></pre><h3 id="mcetoc_1ildb1k5glf">Week 8</h3><p class="align-center"></p><div class="post__iframe"><iframe loading="lazy" width="697" height="392" style="width: 697px; height: 392px;" src="https://www.youtube-nocookie.com/embed/_pPPqxlxh7g?si=Lr_UxUsH6KAmk8Pa" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen="allowfullscreen"></iframe></div>Preliminary animation of human genome uncharacterized proteins. Each frame is an individual sequence with the nucleotides mapped individually in color. (See code below)<p></p><pre class="language-python"><code># from oneball.py (should also copy onto cluster) of VIS 141A
import os
import os.path
import sys
import pygame
from pygame.locals import *
import time
import random
# from mpi4py import MPI
import math
import sqlite3

# comm = MPI.COMM_WORLD
# rank = comm.Get_rank()
# size = comm.Get_size()

# set window position
os.environ['SDL_VIDEO_WINDOW_POS'] = "0,0"
# get this display
#print(dict(os.environ))
os.environ['DISPLAY'] = ':0.0'
# with caffeine installed (sudo apt install caffeine) we can wake the screen
# os.system('caffeinate sleep 1') # passes caffeinate out to shell to wake screensaver

# init clock and display
clock = pygame.time.Clock()
pygame.display.init()
pygame.mouse.set_visible(False)

# get the screen hight and width
disp_info = pygame.display.Info()
width = disp_info.current_w
height = disp_info.current_h
screen_size = (width,height) # adjust screen size manually is monitor is not the same dimensions as the main monitor
# universe_size = (width*2, height*2)
screen = pygame.display.set_mode((screen_size), pygame.NOFRAME)

# sqlite search cross with csv --&gt; get list of uncharacterized protein codes, find each of those in sqlite database
    # wondering if i should just put the protein names into sqlite db as well to better describe the protein and just search that way
    # updated database: genome_human (gene VARCHAR(14), protein VARCHAR(132), sequence VARCHAR(107976), changed_sequence INTEGER)
        # defaults changed_sequence to 0 (represents false), changes to 1 if serial data being read is giving into a change to the sequence (duplicate database or dataset to read)

connection = sqlite3.connect('genome_human.db') # change to full path name?
cursor = connection.cursor()

cursor.execute("SELECT sequence FROM genome_human WHERE protein LIKE '%uncharacterized%'")
# SELECT * will return [(gene, protein, sequence),...]
    # () &lt;- tuple so even if 1 its still (value,) at least
    # obtain thru tuple indexing: seq1 = var[0][0] (1st list item, index 0 of tuple)
var1 = cursor.fetchall()
# cursor.execute("SELECT protein FROM genome_human WHERE protein LIKE 'uncharacterized%'")
# var2 = cursor.fetchall()
# print(len(var1)) #1638 (for protein)
# print(len(var2)) #1564 (for protein)

# check proteins that have augment before "uncharacterized"
# for p in var1:
#     if p not in var2:
#         print(p)
# described uncharacterized:
#     ('LOW QUALITY PROTEIN: uncharacterized protein',)
#     ('putative uncharacterized protein C3orf49',)
# do i want to include those above or even make them smth different based on being putative or low quality?
# print(var)
framerate = 24
runtime = 10 #seconds

def nucleotide_color(nucleotide):
    if nucleotide == "A":
        return 50
    elif nucleotide == "C":
        return 100
    elif nucleotide == "T":
        return 150
    elif nucleotide == "G":
        return 200
    else:
        return 255

def sine_wave(shift,sequence):
    len_sequence = len(sequence)
    for i in range(height):
        # sine transformations -&gt; bsin(ax-w)+h (b increases amplitude, a gives inverse frequency increase, w shifts horizontally, h shifts vertically)
        # pygame.draw.line(screen,(50,100,255),(0,i),((width/2)+((width/2)*math.sin((height/4)*(i-shift))),i)) 
            # moves down the screen
        # pygame.draw.line(screen,(50,100,255),(0,i),((width/2)+((width/2)*math.sin((height/4)*i)),i))
            # this (width/2)+((width/2)*math.sin((height/4)*i)) did NOT do what i expected at all but its GORGEOUS
        if i &gt;= len_sequence:
            pygame.draw.line(screen,(255,255,255),(0,i),((width/2)+((width/2)*math.sin((height/4)*i)),i))
        else:
            pygame.draw.line(screen,(50,100,nucleotide_color(sequence[i])),(0,i),((width/2)+((width/2)*math.sin((height/4)*i)),i))
    return

protein_index = 0
# time to run
# timeToRun = framerate*runtime
timeToRun = len(var1)

for f in range(timeToRun):
    # screen.fill((50,30,255))
    screen.fill((0,0,0))
    sine_wave(f,var1[protein_index][0]) #shift later to go thru each sequence, this would be per frame but i could try other methods later
    pygame.display.update()
    protein_index = protein_index + 1
    if protein_index &gt; len(var1):
        protein_index = 0
    clock.tick(framerate)

connection.close()</code></pre><h3 id="mcetoc_1ildb1k5glg">Week 7</h3><div class="gallery-wrapper"><div class="gallery" data-is-empty="false" data-translation="Add images" data-columns="3"><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250226_124440.jpg" data-size="2801x2172"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/20250226_124440-thumbnail.jpg" alt="" width="720" height="558"></a><figcaption>2x4 screen display setup with vertical screen placement</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups2.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups2-thumbnail.png" alt="" width="720" height="405"></a><figcaption>2x4 vertical screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups-thumbnail.png" alt="" width="720" height="405"></a><figcaption>4x2 horizontal screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups1.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups1-thumbnail.png" alt="" width="720" height="405"></a><figcaption>3x2 horizontal screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups3.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups3-thumbnail.png" alt="" width="720" height="405"></a><figcaption>2x4 horizontal screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups6.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups6-thumbnail.png" alt="" width="720" height="405"></a><figcaption>Split 2x2, 2x2 screen displays</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups5.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups5-thumbnail.png" alt="" width="720" height="405"></a><figcaption>1x8 vertical screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups4.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups4-thumbnail.png" alt="" width="720" height="405"></a><figcaption>4x2 vertical screen display</figcaption></figure><figure class="gallery__item"><a href="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups7.png" data-size="960x540"><img loading="lazy" src="https://n-katty-kat.github.io/ACTinG-for-U/media/posts/13/gallery/vis-160-cluster-setups7-thumbnail.png" alt="" width="720" height="405"></a><figcaption>2 layer shifted 8-screen display</figcaption></figure></div></div><p>Rearranging screen display for cluster: 3x3 horizontal screens to 2x4 vertical screens. Intended to improve scaling and utilization of wall space.</p><p>Fixed genome .fna file reading: Wrote out gene ID, protein name, and sequence to SQLite database. (See code below)</p><p>Fixed PySerial communication and ran sample code to print out simple Serial output from Arduino:</p><ol style="list-style-type: lower-roman;"><li><a href="https://roboticsbackend.com/raspberry-pi-arduino-serial-communication/">https://roboticsbackend.com/raspberry-pi-arduino-serial-communication/</a></li></ol><pre class="language-python"><code># reading the human genome fna file

# code ref from ch 19 https://coding4medicine.com/backup/Python/reading-fasta-files.html
import sys
import time
import sqlite3

f=open("ncbi_dataset/data/GCF_000001405.40/cds_from_genomic.fna", "r") # open file with the genome data

connection = sqlite3.connect('genome_human.db') # change to full path name?
cursor = connection.cursor()

cursor.execute('CREATE TABLE genome_human (gene VARCHAR(14), protein VARCHAR(132), sequence VARCHAR(107976), changed_sequence INTEGER)') #determine the appropriate varchar max for each of these by reading through the file
# https://dba.stackexchange.com/questions/190674/can-a-database-table-not-have-a-primary-key 
    # use rowid to see primary key
# changed_sequence is boolean as integer (0 false, 1 true) --&gt; indicates whether the sequence item has been changed or not by external adjustments to its nucleotides

connection.commit() # commit to creating the file before trying to insert into it

for line in f:
    # read over each line of the file
    if(line.startswith("&gt;")):
        if(new_gene):
            new_gene = False
            sequence_list = sequence_string.split("\n") # remove the \n from btwn each sequence subpart from the fna format and any blank lines
            sequence_string = "" #clear out sequence so that it doesnt concatenate old sequences w new sequences
            sequence_output = "".join(sequence_list) # join together that list to form one continuous sequence line

            cursor.execute('INSERT INTO genome_human VALUES (?,?,?,0);', (gene_id, protein_name, sequence_output))

        identity = line.split(" [") # split the elements of each into 1c1, gene], db], pro], proid], loc], gbkey] (since protein= can have spaces in btwn)
            # gene always gene=###] --&gt; same way to obtain as protein_id with splice or split and cut 5 (after =) to length
            # protein_id will always be protein_id=#####] so can obtain number as split at = and get num minus ending bracket or splice index 11 to length (to keep out bracket)
        # get gene
        for i in identity:
            if(i.startswith("gene=")):
                gene_id = i
                break
        gene_id_len = len(gene_id)
        gene_id = gene_id[5:(gene_id_len-1)] # cuts out "gene=" and "]" ; alternately can put len("gene=") in place of 6

        for i in identity:
            if(i.startswith("protein=")):
                protein_name = i
                break
        protein_name_len = len(protein_name)
        protein_name = protein_name[len("protein="):(protein_name_len-1)]
        
    else:
        new_gene = True
        sequence_string = sequence_string + line # once starting a new gene, the string will be split at "\n" and pieced back together as a string

print("Done")
f.close()
connection.commit()
cursor.close()
connection.close()</code></pre><h3 id="mcetoc_1ildb1k5glg">Week 5</h3><p>Researching FASTA format of human genome data:</p><ol style="list-style-type: lower-roman;"><li><a href="https://coding4medicine.com/backup/Python/reading-fasta-files.html">https://coding4medicine.com/backup/Python/reading-fasta-files.html</a></li><li><a href="https://en.wikipedia.org/wiki/FASTA_format">https://en.wikipedia.org/wiki/FASTA_format</a></li><li><a href="https://www.ncbi.nlm.nih.gov/datasets/taxonomy/9606/">https://www.ncbi.nlm.nih.gov/datasets/taxonomy/9606/</a></li></ol><p>Testing reading out genome data in Python:</p><ol style="list-style-type: lower-roman;"><li>Issue: CSV file ended up being over 9GB!! (from a 2GB FASTA .fna file)</li><li>Resolution: The variable holding each sequence was not being cleared for each new sequence and continuously concatenated the sequences (memory leak...).</li></ol><p>Current issue: PySerial not installing on Raspberry Pi 5 (Spoiler Alert: It was already pre-installed)</p></div></article></main><footer id="copyright"><ul><li>© Jett via Massively</li><li>Text, Photos, Videos, Graphics, and all other non-theme and non-citation materials are creative and intellectual property of A. Jett. None of these materials are permitted for use in AI systems.</li><li>Design: <a href="https://html5up.net" target="_blank" rel="nofollow noopener">HTML5 UP</a></li><li>Powered by Publii</li></ul></footer></div><script src="https://n-katty-kat.github.io/ACTinG-for-U/assets/js/jquery.min.js?v=c9771cc3e90e18f5336eedbd0fffb2cf"></script><script src="https://n-katty-kat.github.io/ACTinG-for-U/assets/js/jquery.scrollex.min.js?v=f89065e3d988006af9791b44561d7c90"></script><script src="https://n-katty-kat.github.io/ACTinG-for-U/assets/js/jquery.scrolly.min.js?v=1ed5a78bde1476875a40f6b9ff44fc14"></script><script src="https://n-katty-kat.github.io/ACTinG-for-U/assets/js/browser.min.js?v=c07298dd19048a8a69ad97e754dfe8d0"></script><script src="https://n-katty-kat.github.io/ACTinG-for-U/assets/js/breakpoints.min.js?v=81a479eb099e3b187613943b085923b8"></script><script src="https://n-katty-kat.github.io/ACTinG-for-U/assets/js/util.min.js?v=4201a626f8c9b614a663b3a1d7d82615"></script><script src="https://n-katty-kat.github.io/ACTinG-for-U/assets/js/main.min.js?v=56233c354bd814758be8bff42f7e13a5"></script><script>/*<![CDATA[*/var images=document.querySelectorAll("img[loading]");for(var i=0;i<images.length;i++){if(images[i].complete){images[i].classList.add("is-loaded")}else{images[i].addEventListener("load",function(){this.classList.add("is-loaded")},false)}};/*]]>*/</script></body></html>